src/components/TodoList.tsx

```diff
@@ -0,0 +1,296 @@
+  setDeleteErrorMessage: React.Dispatch<React.SetStateAction<string>>,
+}
+
+export const TodoList: React.FC<Props> = ({
+  todo,
+  visibleTodos,
+  isLoading,
+  updatingTodoId,
+  tempTodo,
+  updateIndividualTodo,
+  isDoubleClickedName,
+  placeHolderText,
+  setPlaceHolderText,
+  excludedInputRef,
+  isUpdating,
+  setIsDoubleClickedName,
+  deleteTodo,
+  setIsUpdating,
+  setTodos,
+  isPlusOne,
+  isThereIssue,
+  deletedTodoId,
+  isEveryThingDelete,
+  todoStatusChange,
+  toggleFalseTodosId,
+  isEveryThingTrue,
+  setIsThereIssue,
+  setDeleteErrorMessage,
+}) => {
+  const [isTodoRenaming, setIsTodoRenaming] = useState(false);
+  const [reWrittenTodoId, setReWrittenTodoId] = useState(0);
+  const [isTitleSame, setIsTitleSame] = useState(false);
+  const timeoutId = useRef<NodeJS.Timeout | null>(null);
+  const [isModifing, setIsModifing] = useState(false);
+
+  const handleTodoUpdate = async (id: number, newTitle: string) => {
+    const updatedTodo = todo.map((obj) => {
+      if (obj.id === id) {
+        if (obj.title === newTitle) {
+          setIsTitleSame(true);
+        }
+
+        return {
+          ...obj,
+          title: newTitle,
+        };
+      }
+
+      return obj;
+    });
+
+    setReWrittenTodoId(id);
+
+    setTodos(updatedTodo);
+
+    if (newTitle === '') {
+      deleteTodo(id);
+    }
+
+    try {
+      setIsTodoRenaming(true);
+      const todoToUpdate = todo.find((elem) => elem.id === id);
+
+      if (todoToUpdate) {
+        await client.patch(`/todos/${id}`, {
+          ...todoToUpdate,
+          title: newTitle,
+        });
+        setIsUpdating(false);
+      }
+    } catch (error) {
+      setIsThereIssue(true);
+      setDeleteErrorMessage('Unable to update a todo');
+      timeoutId.current = setTimeout(() => {
+        setIsThereIssue(false);
+      }, 3000);
+      throw Error('Unable to update a todo');
+    } finally {
+      setIsTodoRenaming(false);
+      setIsModifing(false);
+    }
+  };
+
+  const handleKeyDown = (
+    event: React.KeyboardEvent<HTMLInputElement>,
+    id: number,
+  ) => {
+    if (event.key === 'Escape') {
+      event.preventDefault();
+      setIsTodoRenaming(false);
+      setIsUpdating(false);
+      setIsModifing(false);
+      setIsDoubleClickedName('');
+    }
+
+    if (placeHolderText === '') {
+      deleteTodo(id);
+    }
+  };
+
+  const handleBlur = (id: number, currentTitle: string) => {
+    if (currentTitle === '' && currentTitle !== tempTodo?.title) {
+      deleteTodo(id);
+    } else if (currentTitle !== tempTodo?.title) {
+      handleTodoUpdate(id, currentTitle);
+    }
+  };
+
+  useEffect(() => {
+    return () => {
+      if (timeoutId.current) {
+        clearTimeout(timeoutId.current);
+      }
+    };
+  }, []);
+
+  return (
+    <section className="todoapp__main">
+      {todo.length !== 0 && (
+        <>
+          {visibleTodos.map((task) => {
+            const shouldChange
```

Comment:
The 'shouldChange' condition is hard to read. Consider breaking it down into separate, well-named variables for better readability.